lib/db/transactions.go

```diff
@@ -58,13 +59,50 @@ func (t readOnlyTransaction) getFileTrunc(key []byte, trunc bool) (FileIntf, boo
 	if err != nil {
 		return nil, false, err
 	}
-	f, err := unmarshalTrunc(bs, trunc)
+	f, err := t.unmarshalTrunc(bs, trunc)
 	if err != nil {
 		return nil, false, err
 	}
 	return f, true, nil
 }
 
+func (t readOnlyTransaction) unmarshalTrunc(bs []byte, trunc bool) (FileIntf, error) {
+	if trunc {
+		var tf FileInfoTruncated
+		err := tf.Unmarshal(bs)
+		if err != nil {
+			return nil, err
+		}
+		return tf, nil
+	}
+
+	var tf protocol.FileInfo
+	if err := tf.Unmarshal(bs); err != nil {
+		return nil, err
+	}
+	if err := t.fillBlockList(&tf); err != nil {
+		return nil, err
+	}
+	return tf, nil
+}
+
+func (t readOnlyTransaction) fillBlockList(fi *protocol.FileInfo) error {
+	if fi.BlockListKey == nil {
+		return nil
+	}
+	bs, err := t.Get(fi.BlockListKey)
+	if err != nil {
+		return err
+	}
+	var bl BlockList
+	if err := bl.Unmarshal(bs); err != nil {
+		return err
+	}
+	fi.Blocks = bl.Blocks
+	fi.BlockListKey = nil
```

Comment:
What the motivation of "niling" `Blocks` and `BlockListKey` in the appropriate place instead of having `protocol.FileInfo` with only the former and another type to be stored with only the latter, preventing any misshaps like forgetting a switch?

Reply:
I was thinking this was simpler and avoids having to copy a bunch of fields to a new type and keep that in sync. As noted I want to remove the *truncated type completely at some point...

Reply:
An alternative could be to have a database-only message type that embeds a fileinfo and has a custom marshaller that nils out the blocks and such. Don’t know if that’s a thing with protobuf, can check.

Reply:
Since the block list anyhow requires special handling (marshalling separately, checking if it's already stored, ...) I don't this would help not forgetting. As is it is hidden away in t.putFile and t.unmarshalTrunc which is where we've historically done the similar thing.